src/internal/log/span.go

```diff
@@ -0,0 +1,138 @@
+package log
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// Level is the level at which to generate Span logs.
+type Level int
+
+const (
+	DebugLevel Level = 1
+	InfoLevel  Level = 2
+	ErrorLevel Level = 3
+)
+
+func (l Level) log(z *zap.Logger, msg string, fields ...Field) {
+	switch l { //exhaustive:enforce
+	case DebugLevel:
+		z.Debug(msg, fields...)
+		return
+	case InfoLevel:
+		z.Info(msg, fields...)
+		return
+	case ErrorLevel:
+		z.Error(msg, fields...)
+		return
+	}
+	z.DPanic("log: internal error: unknown level in Level.log call", zap.Int("level", int(l)))
+	z.Debug(msg, fields...)
+}
+
+// EndSpanFunc is a function that ends a span.
+type EndSpanFunc = func(fields ...Field)
+
+// ErrorL is a Field that marks a span as failed, and logs the end of the span at the provided level.
+func ErrorL(err error, level Level) Field {
+	if err == nil {
+		return zap.Skip()
+	}
+	f := zap.Error(err)
+	f.Integer = int64(level)
+	return f
+}
+
+// ErrorpL is a Field that marks a span as failed if *err is a non-nil error at the time when the
+// span ends.  See ErrorL.
+func ErrorpL(err *error, level Level) Field {
+	// If you're wondering why this ends in "pL" with that capitalization; zap uses the
+	// convention 'lowercase p = pointer', and we use the convention 'uppercase L = level'.
+	f := zap.Reflect("error", err)
+	f.Integer = int64(level)
+	return f
+}
+
+// Errorp is a Field that marks a span as failed.  See ErrorpL.
```

Comment:
Honestly, I expect `Errorp` to be a predicate returning true if its argument is an error, but I get that that’s niche nowadays.

Reply:
Again, this is zap's convention: 

```
$ go doc zap.Field | egrep 'p\('
func Boolp(key string, val *bool) Field
func Complex128p(key string, val *complex128) Field
func Complex64p(key string, val *complex64) Field
func Durationp(key string, val *time.Duration) Field
func Float32p(key string, val *float32) Field
func Float64p(key string, val *float64) Field
func Int16p(key string, val *int16) Field
func Int32p(key string, val *int32) Field
func Int64p(key string, val *int64) Field
func Int8p(key string, val *int8) Field
func Intp(key string, val *int) Field
func Skip() Field
func StackSkip(key string, skip int) Field
func Stringp(key string, val *string) Field
func Timep(key string, val *time.Time) Field
func Uint16p(key string, val *uint16) Field
func Uint32p(key string, val *uint32) Field
func Uint64p(key string, val *uint64) Field
func Uint8p(key string, val *uint8) Field
func Uintp(key string, val *uint) Field
func Uintptrp(key string, val *uintptr) Field
```